File: 2020/Day4/Python/PassportProcessing-Part2.py
import sys
import collections
import itertools
import numpy as np
import re

passports = []

passp = ""
for line in open(f"{sys.path[0]}/../input.txt").read().splitlines():
    if len(line) == 0:
        passports.append(passp)
        passp = ""
    else:
        passp += " " + line.replace("\n"," ")
passports.append(passp)    

needed = ["byr", "iyr", "eyr", "hgt","hcl","ecl","pid"]

def validate(cat, val):
    if cat == "byr":
        return 1920 <= int(val) <= 2002
    elif cat == "iyr":
        return 2010 <= int(val) <= 2020
    elif cat == "eyr":
        return 2020 <= int(val) <= 2030
    elif cat == "hgt":
        unit = val[-2:]
        height = val[:-2]
        if unit == "cm":
            return 150 <= int(height) <= 193
        elif unit == "in":
            return 59 <= int(height) <= 76
    elif cat == "hcl":
        return bool(re.match(r"^#[0-9a-f]{6}$", val))
    elif cat == "ecl":
        return val in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
    elif cat == "pid":
        return bool(re.match(r"^[0-9]{9}$", val))
    elif cat == "cid":
        return True
    return False
ans = 0

passSplit = []

for p in passports:
    for n in needed:
        if n not in p:
            break
    else:
        ps = p.split()
        for i, pss in enumerate(ps):
            ps[i] = pss.split(":")
        passSplit.append(ps)

print(len(passSplit))
   
for p in passSplit:
    for cat, val in p:
        if not validate(cat, val):
            break
    else:
        ans += 1

print(ans)

File: 2020/Day4/Haskell/PassportProcessing-Part2.hs
{-# LANGUAGE OverloadedStrings #-}

import System.IO
import Control.Monad
import qualified Data.Text as Text

printArray:: (Show a, Eq a) => [a] -> String
printArray (x:xs)
    | xs == [] = show x
    | otherwise = (show x) ++ "\n" ++ printArray xs

subsets:: Int -> [a] -> [[a]]
subsets 0 _ = [[]]
subsets _ [] = []
subsets n (x : xs) = map (x :) (subsets (n - 1) xs) ++ subsets n xs

split:: Eq a => a -> [a] -> [[a]]
split d [] = []
split d s = x : split d (drop 1 y) where (x,y) = span (/= d) s

groupPass:: [[Char]] -> [[Char]] -> [[Char]]
groupPass groups remainder
    | (length remainder) == 0 = []
    | otherwise = groups ++ [unwords x] ++ (groupPass [] (drop ((length x) + 1) remainder)) where x = takeWhile(\x -> 0 < (length x)) remainder 

validate:: [[Char]] -> Bool
validate (x:y:xs)
    | x == "byr" = 1920 <= val && val <= 2002 
    | x == "iyr" = 2010 <= val && val <= 2020 
    | x == "eyr" = 2020 <= val && val <= 2030 
    | x == "hgt" && unit == "cm" = 150 <= height && height <= 193
    | x == "hgt" && unit == "in" = 59 <= height && height <= 76
    | x == "hcl" = (length y == 7) && ((head y) == '#') && (all (\n -> elem n ("0123456789abcdef"::String)) (tail y))
    | x == "ecl" = elem y ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
    | x == "pid" = (length y == 9) && (all (\n -> elem n ("0123456789"::String)) y)
    | x == "cid" = True
    | otherwise = False
        where 
            val = read y::Int
            unit = drop ((length y) - 2) y
            height = read (take ((length y) - 2) y)::Int
    
needed:: [Text.Text]
needed = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]

main:: IO()
main = do
    handle <- openFile "../input.txt" ReadMode
    contents <- hGetContents handle
    let rows = (lines contents)
    let filtered = filter (\p -> (all (\n -> Text.isInfixOf n (Text.pack p)) needed)) (groupPass [] rows)
    let splitPass = (map (\s -> (map (\x -> split ':' x) (split ' ' s))) filtered)
    print (length (filter (\p -> (all (\a -> validate a) p)) splitPass))
    hClose handle
